include/seqan3/alphabet/alphabet_base.hpp

```diff
@@ -58,9 +59,15 @@ class alphabet_base
     /*!\name Member types
      * \{
      */
-    //!\brief The char representation; conditional needed to make semi alphabet definitions legal.
+    /*!\brief The char representation; conditional needed to make semi alphabet definitions legal.
```

Comment:
`char_type` is for Alphabet concept, not Semialphabet...
```suggestion
    /*!\brief The char representation; conditional needed to make alphabet definitions legal.
```

Reply:
This is open

Reply:
I added

```
      * \details We need a return type for seqan3::alphabet_base::to_char and seqan3::alphabet_base::assign_char other
      *          than void to make these in-class definitions valid when `char_t` is void.
 ```

I hope this makes it more clear :)

Reply:
I think I understand now: The char type is void for semi alphabets, which makes a conditional necessary. I agree and give green light. 
But what still confuses me: this is alphabet base, not semi alphabet base, so why is `void` allowed at all? And wouldn't it be better to `enable_if` the char type only if it is not void? (The same with assign_char and to_char...) Because we don't want that e.g. `semialphabet::to_char` exists and returns `char`...

Reply:
We don't have a dedicated semi_alphabet_base, our alphabet_base does both cases. If char_t is void it is a semi-alphabet, otherwise it is an alphabet.

> And wouldn't it be better to `enable_if` the char type only if it is not void?

We do a requires on to_char and assign_char if the value is void or not. So to_char is not available if the alphabet_base is a semi-alphabet.

The Problem is more technically, because 

```c++
char_type foo() requires char_type != void
{
   return something;
}
```

has a return statement and if the return type (= char_type) is void, the function definition is invalid and will throw a compiler error, but we want a SFINAE error here, but for that the function definition must be syntactically valid. The crux here is that the return type could be any type, even `void *`, because `foo` will never be considered, because the requires will be false in these cases. 

We could also define it as 

```c++
auto foo() requires char_type != void
{
   return something;
}
```

but we would lose the documentation.

Or define it old-school like:

```c++
std::enable_if_t<char_type != void, char_t> foo() 
{
   return something;
}
```

Another way would be to split out current `alphabet_base` into a `alphabet_base` and a `semi_alphabet_base`, but we have some alphabets (primarily proxies) that can be either conditionally. So we need at some point a mechanism to be either conditionally.

I'm also un-happy that we have a char_type that is not always the same as char_t, but there is no easy way to conditionally not define a type.